automated-testing/ui-testing/readme.md

```diff
@@ -0,0 +1,64 @@
+# UI Testing
+
+This section is primarily geared towards web-based UIs, but the guidance is similar for mobile and OS based applications.  
+
+## Goals
+
+UI testing provides the ability to ensure that users have a consistent visual user experience across a variety of means of access and that the user interaction is consistent with the function requirements.
+
+### UI tests have several goals
+
+- Ensure the UI appearance and interaction satisfy the functional and non-functional requirements
+- Detect changes in the UI both across devices and delivery platforms and between code changes
+- Provide confidence to designers and developers the user experience is consistent
+- Support fast code evolution and refactoring while reducing the risk of regressions
+
+## Evidence and Measures
+
+Integrating UI Tests in to your CI/CD is necessary but more challenging than unit tests.  The increased challenge is that UI tests either need to run headlessly with something like [Puppeteer](https://github.com/puppeteer/puppeteer) or there needs to be more extensive orchestration with Azure DevOps or Github that would handle the full testing integration for you like [BrowserStack](https://www.browserstack.com/automate/azure)
+
+Integrations like `BrowserStack` are nice since they provide Azure DevOps reports as part of the test run.
+
+That said, Azure DevOps supports a variety of test adapters, so you can use any UI Testing framework that supports outputting the test results to one of the output formats listed at [Publish Test Results task](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/publish-test-results?view=azure-devops&tabs=yaml)
+
+## General Guidance
+
+The scope of UI testing should be strategic. UI tests can take a significant amount of time to both implement and run, and it challenging to test every possibility of user interaction in a production application due to the large possible number of possible interactions.
+
+Designing the UI tests around the functional tests makes sense.  For example, given an input form, a UI test would ensure that the visual representation is consistent across devices, is accessible and interactable, and is consistent across code changes.
+
+UI Tests will catch 'runtime' bugs that unit and functional tests won't.  For example if the submit button for an input form is rendered but not clickable due to a positioning bug in the UI, then this could be considered a runtime bug that would not have been caught by unit or functional tests.
+
+UI Tests can run on mock data or snapshots of production data, like in QA or staging.
+
+### Writing Tests
+
+Good UI tests follow a few general principles:
+
+- Choose a UI testing framework that enables quick feedback and is easy to use
+- Design the UI to be easily testable.  For example, add CSS selectors or set the id on elements in a web page to allow easier selecting.
+- Test on all primary devices that the user uses, don't just test on a single device or OS.
```

Comment:
\+ browsers - the nightmares of IE8 still haunts me sometimes 

Do we want to hint at using emulators for devices and browsers?

Reply:
I meant devices to include browsers, the wording is difficult, maybe `mediums` or ` devices and/or browsers`.  Maybe I'll go with the latter as mediums may be too ambiguous